### PR TITLE
Modify frontend autocomplet API to take user input, add caching

### PR DESCRIPTION
**Pull request by alyiwang**

### CODE REVIEW

File: web/app/controllers/api/v1/Search.java [hunk 1]
```diff
@@ -13,60 +13,101 @@
...
+
+    public static Result getSearchAutoComplete() {
+        int size = NumberUtils.toInt(request().getQueryString("size"), DEFAULT_AUTOCOMPLETE_SIZE);
+        String input = request().getQueryString("input");
+
+        String cacheKey = AUTOCOMPLETE_ALL_KEY + (isNotBlank(input) ? "." + input : "");
+        List<String> names = (List<String>) Cache.get(cacheKey);
+        if (names == null || names.size() == 0) {
+            names = SearchDAO.getAutoCompleteList(input, size);
+            Cache.set(cacheKey, names, 3600);
```
Comment (mars-lan) [inline web/app/controllers/api/v1/Search.java]: Use a constant for timeout?

File: web/app/dao/SearchDAO.java [hunk 2]
```diff
@@ -153,80 +155,47 @@
...
+	private final static String GET_FLOW_AUTO_COMPLETE_LIST =
+			"SELECT DISTINCT flow_name FROM flow where flow_name like ? LIMIT ?";
+	private final static String GET_JOB_AUTO_COMPLETE_LIST =
+			"SELECT DISTINCT job_name FROM flow_job where job_name like ? LIMIT ?";
+
+
+	public static List<String> getAutoCompleteList(String input, int limit) {
+		String pattern = isNotBlank(input) ? input + "%" : "%";
+		List<String> names = new ArrayList<>();
+		names.addAll(getJdbcTemplate().queryForList(GET_DATASET_AUTO_COMPLETE_LIST, String.class, pattern, limit));
```
Comment (mars-lan) [inline web/app/dao/SearchDAO.java]: So we can potentially return up to 4*limit number of results?

File: web/app/dao/SearchDAO.java [hunk 3]
```diff
@@ -153,80 +155,47 @@
...
-			List<String> datasetList = getJdbcTemplate().queryForList(GET_DATASET_AUTO_COMPLETE_LIST, String.class);
-			cachedAutoCompleteList =
-					Stream.concat(Stream.concat(datasetList.stream(), metricList.stream()),
-							Stream.concat(flowList.stream(), jobList.stream())).collect(Collectors.toList());
-			Collections.sort(cachedAutoCompleteList);
-			Cache.set(SEARCH_AUTOCOMPLETE_LIST, cachedAutoCompleteList, 60*60);
-		}
-
-		return cachedAutoCompleteList;
+	private final static String SEARCH_AUTOCOMPLETE_LIST = "searchSource";
```
Comment (mars-lan) [inline web/app/dao/SearchDAO.java]: No longer needed?

File: web/app/controllers/api/v1/Search.java (hunk 1 above)
Comment (alyiwang) [inline web/app/controllers/api/v1/Search.java]: Thanks, Fixed

File: web/app/dao/SearchDAO.java (hunk 3 above)
Comment (alyiwang) [inline web/app/dao/SearchDAO.java]: Thanks, fixed

File: web/app/dao/SearchDAO.java (hunk 2 above)
Comment (alyiwang) [inline web/app/dao/SearchDAO.java]: Fixed.

File: web/app/dao/SearchDAO.java [hunk 4]
```diff
@@ -61,7 +63,7 @@
 			" AGAINST ('*$keyword* *v_$keyword*' IN BOOLEAN MODE) ) t " +
 			"ORDER BY rank DESC, `name`, `urn` LIMIT ?, ?;";
 
-	public final static String SEARCH_DATASET_BY_SOURCE_WITH_PAGINATION = "SELECt SQL_CALC_FOUND_ROWS " +
+	private final static String SEARCH_DATASET_BY_SOURCE_WITH_PAGINATION = "SELECt SQL_CALC_FOUND_ROWS " +
```
Comment (theseyi) [inline web/app/dao/SearchDAO.java]: carried over from previous impl, lowercase `t` in `SELECt`

File: web/app/dao/SearchDAO.java [hunk 5]
```diff
@@ -153,80 +155,60 @@
...
-	{
-		List<String> cachedAutoCompleteListForDataset = (List<String>)Cache.get(SEARCH_AUTOCOMPLETE_LIST_DATASET);
-		if (cachedAutoCompleteListForDataset == null || cachedAutoCompleteListForDataset.size() == 0)
-		{
-			List<String> datasetList = getJdbcTemplate().queryForList(GET_DATASET_AUTO_COMPLETE_LIST, String.class);
-			cachedAutoCompleteListForDataset = datasetList.stream().collect(Collectors.toList());
-			Collections.sort(cachedAutoCompleteListForDataset);
-			Cache.set(SEARCH_AUTOCOMPLETE_LIST_DATASET, cachedAutoCompleteListForDataset, 60*60);
+	public static List<String> getAutoCompleteListDataset(String input, int limit) {
+		String pattern = isNotBlank(input) ? input + "%" : "%";
```
Comment (theseyi) [inline web/app/dao/SearchDAO.java]: So I guess we are only doing a prefix match for now?

**Pull request merged**